src/main/java/io/cdap/plugin/gcp/gcs/sink/GCSOutputFormatProvider.java

```diff
@@ -0,0 +1,246 @@
+      InterruptedException {
+      OutputCommitter delegateCommitter = getDelegateFormatInstance(taskAttemptContext.getConfiguration())
+        .getOutputCommitter(taskAttemptContext);
+      return new GCSOutputCommitter(delegateCommitter);
+    }
+  }
+
+  /**
+   * OutputCommitter for GCS
+   */
+  public static class GCSOutputCommitter extends OutputCommitter {
+
+    private final OutputCommitter delegate;
+
+    public GCSOutputCommitter(OutputCommitter delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public void setupJob(JobContext jobContext) throws IOException {
+      delegate.setupJob(jobContext);
+    }
+
+    @Override
+    public void cleanupJob(JobContext jobContext) throws IOException {
+      delegate.cleanupJob(jobContext);
+    }
+
+    @Override
+    public void commitJob(JobContext jobContext) throws IOException {
+      delegate.commitJob(jobContext);
+    }
+
+    @Override
+    public void abortJob(JobContext jobContext, JobStatus.State state) throws IOException {
+      delegate.abortJob(jobContext, state);
+    }
+
+    @Override
+    public void setupTask(TaskAttemptContext taskAttemptContext) throws IOException {
+      delegate.setupTask(taskAttemptContext);
+    }
+
+    @Override
+    public boolean needsTaskCommit(TaskAttemptContext taskAttemptContext) throws IOException {
+      return delegate.needsTaskCommit(taskAttemptContext);
+    }
+
+    @Override
+    public void commitTask(TaskAttemptContext taskAttemptContext) throws IOException {
+      //Update the metric before committing the file.
+      try {
+        updateMetricMetaData(taskAttemptContext);
+      } catch (Exception exception) {
+        LOG.warn("Unable to record metric for task. Metric emitted for the number of affected rows may be incorrect.",
+                 exception);
+      }
+
+      delegate.commitTask(taskAttemptContext);
+    }
+
+    private void updateMetricMetaData(TaskAttemptContext taskAttemptContext) throws IOException {
+      if (!(delegate instanceof FileOutputCommitter)) {
+        return;
+      }
+
+      FileOutputCommitter fileOutputCommitter = (FileOutputCommitter) delegate;
+      Configuration configuration = taskAttemptContext.getConfiguration();
+      //Task is not yet committed, so should be available in attempt path
+      Path taskAttemptPath = fileOutputCommitter.getTaskAttemptPath(taskAttemptContext);
+      if (configuration == null || taskAttemptPath == null) {
+        return;
+      }
+
+      //read the count from configuration
+      String keyInConfig = String.format(RECORD_COUNT_FORMAT, taskAttemptContext.getTaskAttemptID());
+      Map<String, String> metaData = new HashMap<>();
+      metaData.put(GCSBatchSink.RECORD_COUNT, String.valueOf(configuration.getLong(keyInConfig, 0L)));
+      StorageClient storageClient = getStorageClient(configuration);
+      //update metadata on any file - typically only one will be there
```

Comment:
typically, or always? is the taskAttemptPath a file or a directory? If it's not always a single blob, wouldn't the metrics end up incorrect?

Reply:
taskAttemptPath is a directory. The exact blob name depends on the format. Yes in this case the expectation is to find a unique file because each task attempt path contains the task Id in it and record writer for that task creates a unique file with task Id as well. I can change the comment to reflect that.